src/censored.jl

```diff
@@ -0,0 +1,168 @@
+isupperbounded(d::Censored) = isupperbounded(d.uncensored) || isfinite(d.upper)
+
+minimum(d::Censored) = max(minimum(d.uncensored), d.lower)
+maximum(d::Censored) = min(maximum(d.uncensored), d.upper)
+
+function insupport(d::Censored{<:UnivariateDistribution}, x::Real)
+    return d.lower <= x <= d.upper && insupport(d.uncensored, x)
+end
+
+### evaluation
+
+quantile(d::Censored, p::Real) = clamp(quantile(d.uncensored, p), d.lower, d.upper)
+
+function pdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if x == lower
+        cdf(d0, x)
+    elseif x == upper
+        uccdf = ccdf(d0, x)
+        if value_support(typeof(d)) === Discrete
+            uccdf - pdf(d0, x)
+        else
+            uccdf
+        end
+    else
+        result = pdf(d0, x)
+        lower < x < upper ? result : zero(result)
+    end
+end
+
+function logpdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if x == lower
+        logcdf(d0, x)
+    elseif x == upper
+        ulogccdf = logccdf(d0, x)
+        if value_support(typeof(d)) === Discrete
+            logsubexp(ulogccdf, logpdf(d0, x))
+        else
+            ulogccdf
+        end
+    else
+        result = logpdf(d0, x)
+        lower < x < upper ? result : oftype(result, -Inf)
+    end
+end
+
+function cdf(d::Censored, x::Real)
+    result = cdf(d.uncensored, x)
+    return if x < d.lower
+        zero(result)
+    elseif x < d.upper
+        result
+    else
+        one(result)
+    end
+end
+
+function logcdf(d::Censored, x::Real)
+    result = logcdf(d.uncensored, x)
+    return if x < d.lower
+        oftype(result, -Inf)
+    elseif x < d.upper
+        result
+    else
+        zero(result)
+    end
+end
+
+function ccdf(d::Censored, x::Real)
+    result = ccdf(d.uncensored, x)
+    return if x > d.upper
+        zero(result)
+    elseif x > d.lower
+        result
+    else
+        one(result)
+    end
+end
+
+function logccdf(d::Censored, x::Real)
+    result = logccdf(d.uncensored, x)
+    return if x > d.upper
+        oftype(result, -Inf)
+    elseif x > d.lower
+        result
+    else
+        zero(result)
+    end
+end
+
+
+## random number generation
+
+function rand(rng::AbstractRNG, d::Censored)
+    return clamp(rand(rng, d.uncensored), d.lower, d.upper)
+end
+
+
+## show
+
+function show(io::IO, d::Censored)
+    print(io, "Censored(")
+    d0 = d.uncensored
+    uml, namevals = _use_multline_show(d0)
+    uml ? show_multline(io, d0, namevals) :
+          show_oneline(io, d0, namevals)
+    print(io, ", range=(", d.lower, ", ", d.upper, ")")
+    uml && println(io)
+end
+
+_use_multline_show(d::Censored) = _use_multline_show(d.uncensored)
+
+
+#### statistics
+
+quantile(d::Censored, p::Real) = clamp(quantile(d.uncensored, p), d.lower, d.upper)
+
+median(d::Censored) = clamp(median(d.uncensored), d.lower, d.upper)
```

Comment:
Need to check this does the right thing with discrete distributions where the return value is the average of upper and lower median.

Reply:
Would one check this by averaging the medians computed from many samples? If so, looks fine when the bounds is in the support of the uncensored distribution, but not quite right when it's outside the support:

```julia
julia> d0 = DiscreteUniform(1, 4)
DiscreteUniform(a=1, b=4)

julia> median(d0)
2.5

julia> median(censored(d0, 1, 3))
2.5

julia> mean_and_std(median(rand(censored(d0, 1, 3), 1000)) for _ in 1:10000)  # looks fine
(2.50355, 0.4932403228666829)

julia> median(censored(d0, 1, 2))
2.0

julia> mean_and_std(median(rand(censored(d0, 1, 2), 1000)) for _ in 1:10000)  # also fine
(2.0, 0.0)

julia> median(censored(d0, 1, 2.8))
2.5

julia> mean_and_std(median(rand(censored(d0, 1, 2.8), 1000)) for _ in 1:10000)  # whoops
(2.3993599999997923, 0.39430838173945837)

julia> median(censored(d0, 1, 2.3))
2.3

julia> mean_and_std(median(rand(censored(d0, 1, 2.3), 1000)) for _ in 1:10000)  # whoops
(2.1517849999998035, 0.14818046824771816)

julia> mean_and_std(quantile(rand(censored(d0, 1, 2.3), 1000), 0.5) for _ in 1:10000)  # quantile has the same problem
(2.152714999999804, 0.14814940718641226)
```

Reply:
@mschauer as far as I can tell, `DiscreteUniform` is the only distribution in Distributions that behaves this way. All other discrete distributions with integer support take care to return an integer:
```
julia> median(Binomial(10, 0.55))
6

julia> median(Binomial(10, 0.55 - eps()))
5
```

Were there other examples you were thinking of that do this? And is this desired behavior for `median`?

Reply:
```
  median(d::UnivariateDistribution)

  Return the median value of distribution d.
```
Looks like we don't have a conscious rule?

Reply:
Perhaps not a conscious rule, but it seems an implicit one, with one violation. From surveying our discrete distributions:

Approach 1: smallest point in support s.t. cdf ≥ 1/2 (when exists)
- `Bernoulli(0.5)`
- `Categorical([0.2, 0.3, 0.0, 0.5])`
- `DiscreteNonParametric(1:4, [0.2, 0.3, 0.0, 0.5])` (smallest index, not point)
- `Geometric(1 - sqrt(1//2))`
- `Hypergeometric(1, 1, 1)`
- `Poisson(1.6783469900167)`

Approach 2: point in support s.t. cdf == 1/2 or mean of two such points
- `DiscreteUniform(1, 4)`

Didn't look at:
- `NegativeBinomial`
- `PoissonBinomial`
- `Skellam`
- `Soliton`

Reply:
I agree. Probably only `DiscreteUniform` needs to be fixed. In that case, the implementation here is already correct.

Reply:
https://github.com/JuliaStats/Distributions.jl/issues/1474

Reply:
#1473 haha, beat you to it

Reply:
@sethaxen `DiscreteUniform` is fixed on the master branch :slightly_smiling_face:

Reply:
Awesome! Thanks!